command/agent/fs_endpoint.go

```diff
@@ -916,7 +932,21 @@ func (s *HTTPServer) logs(follow, plain bool, offset int64,
 			return nil
 		}
 
-		//Since we successfully streamed, update the overall offset/idx.
+		// defensively check to make sure StreamFramer hasn't stopped
+		// running to avoid tight loops with goroutine leaks as in
+		// #3342
+		select {
+		case <-framer.ExitCh():
+			err := parseFramerErr(framer.Err())
+			if err == syscall.EPIPE {
```

Comment:
Can this logic be in parseFramerErr instead, since it already has logic related to translating errors? I.e, does the caller need to do the translation of `syscall.EPIPE` -> nil error directly?

Reply:
it returns the error from parseFramerErr in line 740, but has this extra logic here where it turns that to nil. Good question though - why do we treat the error two different ways?

Reply:
Reading this again, I am not convinced this block is necessary (saw your comment about defensive code). But if its here, I would make it return the value of parseFramerErr rather than `nil`.

Reply:
This change makes more sense in the actual code and not in the code review I promise. :) The code review cuts out the code 10 lines up that follow similar return behavior when handling the error from `s.stream()`

The error returned from `s.stream()` is already passed through `parseFramerErr()`, then line 924 converts EPIPE to nil in this func.

So this defensive code converts the framer error with the helper func and then follows the same `if err == EPIPE { return nil }` behavior as above.

Sorry the code is obtuse but the good news is that it only determines whether or not an error is logged when a client disconnects (EPIPE errors are *not* logged as they're expected).

The important thing in this defensive code is that we return from this infinite `for{}` loop if the framer has exited. The error code only matters for logging.